HerPublicWebsite.BusinessLogic/ExternalServices/OsPlaces/OsPlacesApi.cs

```diff
@@ -66,12 +68,19 @@ public async Task<List<Address>> GetAddressesAsync(string postcode, string build
 
             var joinedAddresses = dpaAddresses.Concat(lpiAddresses.Where(la => !dpaUprns.Contains(la.Uprn))).ToList();
 
-            var filteredResults = buildingNameOrNumber is null
+            // Filter by the building name or number the user provided.
+            // Some users may fill in the full street address (e.g. 45 High Street), we'd like to match that if possible
+            var userFilterParts = NonAlphanumericCharacters
+                .Replace((buildingNameOrNumber ?? "").ToLower(), " ")
+                .Split(' ');
+
+            var filteredResults = userFilterParts.Length == 0
                 ? joinedAddresses
-                : joinedAddresses.Where(a =>
-                    a.AddressLine1.ToLower().Contains(buildingNameOrNumber.ToLower())
-                    || a.AddressLine2.ToLower().Contains(buildingNameOrNumber.ToLower()))
-                .ToList();
+                : joinedAddresses.Where(a => 
+                        userFilterParts.All(userFilterPart =>
+                            a.AddressLine1.ToLower().Contains(userFilterPart) ||
+                            a.AddressLine2.ToLower().Contains(userFilterPart)))
+                    .ToList();
```

Comment:
So the logic is basically: break the provided string into tokens (e.g. '45', 'High', 'Street') and make sure each one matches some part of the address?

How does this handle substrings - e.g. if I put '1', it'll match 'Flat 1', but will it also match 'Flat 10'?

Erring on the side of matching feels good, so potentially better than trying to get more exact matches - but perhaps worth flagging.

Reply:
That's right. We already matched "1" with "10" before this change too. As you say, it's safer and the downside is pretty small.